### PR TITLE
Fail queries if workflow is terminated without WorkflowTaskStarted event

### PR DESCRIPTION
**Pull request by prathyushpv**
**What changed?**
Added condition to fail query if workflow has terminated and it did not have WorkflowTaskStarted event.

**Why?**
If the workflow never had WorkflowTaskStarted event, query will eventually fail without a clear error.
Adding a clear error message in this case.

**How did you test it?**
Unit tests.

**Potential risks**
None

**Is hotfix candidate?**
No

### CODE REVIEW

**Comment by bergundy** [comment]
Can you explain why you eagerly fail the query? What's the motivation for this?
I'm not sure how the server and all of the SDKs handle this case but technically it should be possible to query workflows even if they've not been started yet in SDKs that allow registering query handlers statically.

**Comment by prathyushpv** [comment]
> Can you explain why you eagerly fail the query? What's the motivation for this? I'm not sure how the server and all of the SDKs handle this case but technically it should be possible to query workflows even if they've not been started yet in SDKs that allow registering query handlers statically.



> 
In this case, the query will fail with the error "Error: 4 DEADLINE_EXCEEDED: context deadline exceeded". It will be more useful in that case to get a clear error message.

Is it possible to set a query handler even before workflow starts executing?

**Comment by bergundy** [comment]
> At this point, history service will wait for the result of the query in a workflow task response or until it is safe to despatch the query through matcher service. Both of these conditions will never become true. A workflow task has to be finished for any of these conditions to become true. So the request will eventually timeout

Why can’t history dispatch the task though?
Not sure it should, just wondering why that is.

File: service/history/api/queryworkflow/api.go [hunk 1]
```diff
@@ -123,6 +123,12 @@ func Invoke(
 		return nil, consts.ErrWorkflowTaskNotScheduled
 	}
 
+	if _, status := mutableState.GetWorkflowStateStatus(); status == enumspb.WORKFLOW_EXECUTION_STATUS_TERMINATED &&
```
Comment (yiminc) [inline service/history/api/queryworkflow/api.go:126]: Timeout or other closed status as well. Should check status != WORKFLOW_EXECUTION_STATUS_RUNNING

File: service/history/consts/const.go [hunk 2]
```diff
@@ -103,6 +103,8 @@ var (
 	ErrResourceExhaustedBusyWorkflow = serviceerror.NewResourceExhausted(enums.RESOURCE_EXHAUSTED_CAUSE_BUSY_WORKFLOW, "Workflow is busy.")
 	// ErrResourceExhaustedAPSLimit is an error indicating user has reached their action per second limit
 	ErrResourceExhaustedAPSLimit = serviceerror.NewResourceExhausted(enums.RESOURCE_EXHAUSTED_CAUSE_APS_LIMIT, "Action per second limit exceeded.")
+	// ErrWorkflowTerminatedBeforeWorkflowTaskStarted is an error indicating workflow execution was terminated before WorkflowTaskStarted event
+	ErrWorkflowTerminatedBeforeWorkflowTaskStarted = serviceerror.NewWorkflowNotReady("Workflow execution terminated before WorkflowTaskStarted event")
```
Comment (yiminc) [inline service/history/consts/const.go:107]: replace terminated to closed

**Review by yiminc — CHANGES_REQUESTED** [review]

**Comment by yiminc** [comment]
> > At this point, history service will wait for the result of the query in a workflow task response or until it is safe to despatch the query through matcher service. Both of these conditions will never become true. A workflow task has to be finished for any of these conditions to become true. So the request will eventually timeout
> 
> Why can’t history dispatch the task though? Not sure it should, just wondering why that is.

I think some SDK will panic on query task if workflow is not started.

File: service/history/api/queryworkflow/api.go (hunk 1 above)
Comment (prathyushpv) [reply service/history/api/queryworkflow/api.go]: Done

File: service/history/consts/const.go (hunk 2 above)
Comment (prathyushpv) [reply service/history/consts/const.go]: Done

File: service/history/api/queryworkflow/api.go (hunk 1 above)
Comment (yycptt) [reply service/history/api/queryworkflow/api.go]: Use mutableState.IsWorkflowExecutionRunning() ?

File: service/history/api/queryworkflow/api.go [hunk 3]
```diff
@@ -123,6 +123,12 @@ func Invoke(
 		return nil, consts.ErrWorkflowTaskNotScheduled
 	}
 
+	if _, status := mutableState.GetWorkflowStateStatus(); status != enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING &&
+		mutableState.GetLastWorkflowTaskStartedEventID() == common.EmptyEventID {
```
Comment (yycptt) [inline service/history/api/queryworkflow/api.go:127]: mutableState.HadOrHasWorkflowTask() ?

Comment (prathyushpv) [reply service/history/api/queryworkflow/api.go:127]: Yeah!

File: service/history/api/queryworkflow/api.go (hunk 1 above)
Comment (prathyushpv) [reply service/history/api/queryworkflow/api.go]: IsWorkflowExecutionRunning() checks `ms.executionState.State`. 
Here I am checking `ms.executionState.Status`. 
is checking `WORKFLOW_EXECUTION_STATE_RUNNING` or `WORKFLOW_EXECUTION_STATE_RUNNING` the same thing in this case?

Comment (prathyushpv) [reply service/history/api/queryworkflow/api.go]: It makes sense to use mutableState.IsWorkflowExecutionRunning(). Updated.

**Review by yiminc — APPROVED** [review]

**Pull request merged**